whylog/config/rule.py

```diff
@@ -57,6 +65,265 @@ def get_causes_parsers(self):
+            5. Pop effect parser search range.
+        """
+        parser_ranges = {
+            self.EFFECT_NUMBER: self._get_effect_range(effect_clues, group, group_type)
+        }
+        queue = deque([self.EFFECT_NUMBER])
+        aggregated_constraints = self._aggregate_constraints()
+        used_parsers = set([self.EFFECT_NUMBER])
+        while queue:
+            parser_number = queue.popleft()
+            for constraint in aggregated_constraints[parser_number]:
+                clues_groups = constraint['clues_groups']
+                depended_parser_number = clues_groups[0][0]
+                base_parser_number = clues_groups[1][0]
+                if depended_parser_number in used_parsers:
+                    continue
+                if not self._is_primary_key_constraint(clues_groups):
+                    continue
+                _, group_type = self._causes[depended_parser_number - 1].get_primary_key_group()
+                parser_ranges[depended_parser_number] = self._calculate_parser_bounds(
+                    base_parser_number, constraint['params'], group_type, parser_ranges
+                )
+                queue.append(depended_parser_number)
+                used_parsers.add(depended_parser_number)
+        self.create_ranges_for_unconnected_parsers(effect_clues, group, group_type, parser_ranges)
+        parser_ranges.pop(self.EFFECT_NUMBER)
+        return parser_ranges
+
+    def _get_effect_range(self, effect_clues, group_number, group_type):
+        """
+        This method basing on effect clues and effect primary key creates effect
+        search range for effect primary key type which in LEFT_BOUND = RIGHT_BOUND and equals
+        primary key value from effect clues.
+        This method is invokes when algorithm with calculate parsers ranges starts. It give
+        set up for this algorithm.
+        Example:
+            clues_groups = {'effect' : Clue((11, 'aaa'), '11 aaa', LineSource('foo', 'bar'))}
+            group_number = 1
+            group_type = 'int'
+            expected returned value : {EFFECT_NUMBER: {'int': {LEFT_BOUND: 11, RIGHT_BOUND: 11}}}
+        """
+        # Here assumption that len of primary_keys_groups equals 1
+        primary_group_value = effect_clues[self.get_effect_name()].regex_parameters[group_number - 1] # yapf: disable
+        return {
+            group_type: {
+                InvestigationStep.LEFT_BOUND: primary_group_value,
+                InvestigationStep.RIGHT_BOUND: primary_group_value
+            }
+        }
+
+    def _aggregate_constraints(self):
+        """
+        This method groups delta constraints by equal base parser
+        """
+        parser_with_constraints = defaultdict(list)
+        for constraint in self._constraints:
+            if constraint['name'] in self.DELTA_CONSTRAINTS:
+                base_parser = constraint['clues_groups'][1][0]
```

Comment:
Now I know the explanation why you need exactly `constraint['clues_groups'][1][0]`. It it because of specific of so called "delta constraints" - parser no 1 is a base parser, that corresponds to the effect, and parser no 0 is a depended parser, which corresponds to the cause - am I right? So if it is explained somewhere in docstrings, someone, who will read this function later, will be confused because the numbers 1 and 0 would be mysterious and he will not know where to look for the explanation. So I suggest you to make special functions for base and dependend parsers, that returns `constraint['clues_groups'][1][0]` (and analogously for dependend parser).